logger/analyticslogger.go

```diff
@@ -0,0 +1,198 @@
+package logger
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/firehose"
+	"github.com/aws/aws-sdk-go/service/firehose/firehoseiface"
+	"github.com/eapache/go-resiliency/retrier"
+)
+
+//go:generate mockgen -package logger -destination mock_firehose.go github.com/aws/aws-sdk-go/service/firehose/firehoseiface FirehoseAPI
+
+// AnalyticsLogger writes to Firehose instead of the logging pipeline
+type AnalyticsLogger struct {
+	KayveeLogger
+	fhStream        string
+	fhAPI           firehoseiface.FirehoseAPI
+	batch           []*firehose.Record
+	batchBytes      int
+	maxBatchRecords int
+	maxBatchBytes   int
+	mu              sync.Mutex
+}
+
+var _ KayveeLogger = &AnalyticsLogger{}
+var _ io.WriteCloser = &AnalyticsLogger{}
+
+// firehosePutRecordBatchMaxRecords is an AWS limit.
+// https://docs.aws.amazon.com/firehose/latest/APIReference/API_PutRecordBatch.html
+const firehosePutRecordBatchMaxRecords = 500
+
+// firehosePutRecordBatchMaxBytes is an AWS limit on total bytes in a PutRecordBatch request.
+// https://docs.aws.amazon.com/firehose/latest/APIReference/API_PutRecordBatch.html
+const firehosePutRecordBatchMaxBytes = 4000000
+
+// AnalyticsLoggerConfig configures things related to collecting analytics.
+type AnalyticsLoggerConfig struct {
+	// Environment is the name of the environment to point to. Default is _DEPLOY_ENV.
+	Environment string
+	// DBName is the name of the ark db.
+	DBName string
+	// Region is the region where this is running. Defaults to _POD_REGION.
+	Region string
+	// FirehosePutRecordBatchMaxRecords overrides the default value (500) for the maximum number of records to send in a firehose batch.
+	FirehosePutRecordBatchMaxRecords int
+	// FirehosePutRecordBatchMaxBytes overrides the default value (4000000) for the maximum number of bytes to send in a firehose batch.
+	FirehosePutRecordBatchMaxBytes int
+	// FirehoseAPI defaults to an API object configured with Region, but can be overriden here.
+	FirehoseAPI firehoseiface.FirehoseAPI
+}
+
+// NewAnalyticsLogger returns a logger that writes to an analytics ark db.
+// It takes as input the db name and the ark db config file.
+func NewAnalyticsLogger(alc AnalyticsLoggerConfig) (*AnalyticsLogger, error) {
+	l := New(alc.DBName)
+	al := &AnalyticsLogger{KayveeLogger: l}
+	l.SetOutput(al)
+	env, dbname := alc.Environment, alc.DBName
+	if env == "" {
+		env = os.Getenv("_DEPLOY_ENV")
+	}
+	if env == "" {
+		return nil, errors.New("_DEPLOY_ENV not set")
```

Comment:
placement is a little confusing 🤔. we can move it into the closure above or actually return something closer to `env could not be set`